datafusion/src/optimizer/timestamp_evaluation.rs

```diff
@@ -0,0 +1,160 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Optimizer rule to replace timestamp expressions to constants.
+//! This saves time in planning and executing the query.
+use crate::error::Result;
+use crate::logical_plan::{Expr, LogicalPlan};
+use crate::optimizer::optimizer::OptimizerRule;
+
+use super::utils;
+use crate::physical_plan::functions::BuiltinScalarFunction;
+use crate::scalar::ScalarValue;
+use chrono::{DateTime, Utc};
+
+/// Optimization rule that replaces timestamp expressions with their values evaluated
+pub struct TimestampEvaluation {
+    timestamp: DateTime<Utc>,
```

Comment:
@alamb Do you think this approach  is correct? I have my doubts as we might create an optimizer and reuse across multiple queries.

Reply:
Now timestamp is initialised during the actual optimize call

Reply:
I think this is a clever approach, indeed the execution context with optimization can be reused across queries.
Also, one could disable the optimization rule (or execute an optimixedy plan). The result of the query needs to be correct in those cases too.

My suggestion would be to add a field to the `ExecutionContext` at the start of a query pass it to the optimizers / evaluation code.